javaslang/src/main/java/javaslang/collection/ArrayType.java

```diff
@@ -0,0 +1,147 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.io.Serializable;
+
+abstract class ArrayType<T> implements Serializable {
```

Comment:
Mostly done - and pushed -, not sure about embedding them to the same file, though.
What's the advantage exactly, are we trying to simulate a sealed trait in Java this way?

Pushed most of the changes, if it's easier for you to merge-and-modify rather than explain further, I won't mind :)